### PR TITLE
[ROCm] route fp16 backward GEMMs to rocBLAS to preserve subnormals

### LINKED ISSUE #182952: hipBLASLt path ignores backward fp16_alt_impl for Half GEMMs, causing FTZ in backward and silent training divergence on gfx90a
### 🐛 Describe the bug

# Summary

On AMD gfx90a (MI250X), PyTorch 2.7+ defaults `blasPreferredBackend()` to `Cublaslt` on ROCm. On this path, backward `float16` GEMMs appear to run without the ROCm backward-pass alternative implementation that preserves fp16 subnormals. As a result, subnormal fp16 inputs in backward GEMMs are flushed to zero, which can silently corrupt gradients and cause training divergence.

Forcing rocBLAS with:

```python
torch.backends.cuda.preferred_blas_library("cublas")
```

restores the expected behavior on the same system.

## Minimal reproducer

```python
import torch

dtype = torch.float16
M = K = N = 64
sub_val = torch.finfo(dtype).tiny / 2
ref = M * sub_val

def check_backward(force_rocblas=False):
    if force_rocblas:
        torch.backends.cuda.preferred_blas_library("cublas")

    print("PyTorch:", torch.__version__)
    print("ROCm:", torch.version.hip)
    if hasattr(torch.backends.cuda, "preferred_blas_library"):
        print("preferred_blas_library:", torch.backends.cuda.preferred_blas_library())

    x = torch.ones(M, K, dtype=dtype, device="cuda")
    w = torch.nn.Parameter(torch.ones(K, N, dtype=dtype, device="cuda"))
    d = torch.full((M, N), sub_val, dtype=dtype, device="cuda")

    (x @ w).backward(d)
    torch.cuda.synchronize()
    val = w.grad[0, 0].item()
    print("grad[0,0] =", val, "expected =", ref)

print("Default backend")
check_backward(force_rocblas=False)

print("\nForced rocBLAS")
check_backward(force_rocblas=True)
```

## Observed behavior

- PyTorch 2.3.0 + ROCm 6.0: default backend is rocBLAS, backward result matches `ref`.
- PyTorch 2.9.1 + ROCm 6.4: default backend is hipBLASLt, backward result is `0.0`.
- PyTorch 2.9.1 + ROCm 6.4 with forced rocBLAS: backward result again matches `ref`.

With BLAS logging enabled, rocBLAS backward GEMMs are dispatched with `flags 4` (`fp16_alt_impl`), while hipBLASLt forward/backward matmuls look identical apart from transpose, suggesting the backward guard is not wired through the hipBLASLt path.

## Expected behavior

One of these should hold:

1. Backward fp16 GEMMs on ROCm should preserve the current rocBLAS behavior when PyTorch selects hipBLASLt by default.
2. If hipBLASLt cannot support the backward-safe path yet, gfx90a should not default to `Cublaslt` for fp16 training workloads where this changes numerical behavior.

## Likely root cause

`ROCmBackwardPassGuard` is honored on the rocBLAS path, where PyTorch passes `rocblas_gemm_flags_fp16_alt_impl`, but not on `gemm_internal_cublaslt<at::Half>()`. The [PyTorch numerical accuracy notes](https://pytorch.org/docs/stable/notes/numerical_accuracy.html) state that PyTorch uses alternate implementations during backward for FP16 denorm handling, but that no longer seems true on gfx90a once the default backend switched to hipBLASLt in 2.7.0.

### Versions

Pytorch 2.3.0 + ROCm6.0
```
PyTorch version: 2.3.0+rocm6.0
Is debug build: False
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: 6.0.32830-d62f6a171

OS: SUSE Linux Enterprise Server 15 SP5 (x86_64)
GCC version: Could not collect
Clang version: 17.0.0 (https://github.com/RadeonOpenCompute/llvm-project roc-6.0.3 24012 af27734ed982b52a9f1be0f035ac91726fc697e4)
CMake version: version 4.1.2
Libc version: glibc-2.31

Python version: 3.12.12 | packaged by conda-forge | (main, Oct 13 2025, 14:34:15) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.4.0-150600.23.73_15.0.14-cray_shasta_c-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: AMD Instinct MI250X (gfx90a:sramecc+:xnack-)
Nvidia driver version: Could not collect
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: 6.0.32830
MIOpen runtime version: 3.0.0
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           48 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  128
On-line CPU(s) list:                     0-127
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 7A53 64-Core Processor
CPU family:                              25
Model:                                   48
Thread(s) per core:                      2
Core(s) per socket:                      64
Socket(s):                               1
Stepping:                                1
Frequency boost:                         enabled
CPU max MHz:                             3541.0149
CPU min MHz:                             1500.0000
BogoMIPS:                                3992.69
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm ibpb_exit_to_user
Virtualization:                          AMD-V
L1d cache:                               2 MiB (64 instances)
L1i cache:                               2 MiB (64 instances)
L2 cache:                                32 MiB (64 instances)
L3 cache:                                256 MiB (8 instances)
NUMA node(s):                            4
NUMA node0 CPU(s):                       0-15,64-79
NUMA node1 CPU(s):                       16-31,80-95
NUMA node2 CPU(s):                       32-47,96-111
NUMA node3 CPU(s):                       48-63,112-127
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; Safe RET
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable: No microcode
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] pytorch-lightning==2.5.5
[pip3] pytorch-triton-rocm==3.0.0
[pip3] torch==2.3.0+rocm6.0
[pip3] torch-geometric==2.6.1
[pip3] torchaudio==2.3.0+rocm6.0
[pip3] torchinfo==1.8.0
[pip3] torchmetrics==1.8.2
[pip3] torchvision==0.18.0+rocm6.0
[conda] numpy                                 1.26.4               pypi_0              pypi
[conda] pytorch-lightning                     2.5.5                pypi_0              pypi
[conda] pytorch-triton-rocm                   3.0.0                pypi_0              pypi
[conda] torch                                 2.3.0+rocm6.0        pypi_0              pypi
[conda] torch-geometric                       2.6.1                pypi_0              pypi
[conda] torchaudio                            2.3.0+rocm6.0        pypi_0              pypi
[conda] torchinfo                             1.8.0                pypi_0              pypi
[conda] torchmetrics                          1.8.2                pypi_0              pypi
[conda] torchvision                           0.18.0+rocm6.0       pypi_0 
```
Pytorch2.9.1+Rocm6.4
```
PyTorch version: 2.9.1+rocm6.4
Is debug build: False
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: 6.4.43484-123eb5128

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.12.3 (main, Jan 22 2026, 20:57:42) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.4.0-150600.23.73_15.0.14-cray_shasta_c-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: AMD Instinct MI250X (gfx90a:sramecc+:xnack-)
Nvidia driver version: Could not collect
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: 6.4.43484
MIOpen runtime version: 3.4.0
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           48 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  128
On-line CPU(s) list:                     0-127
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 7A53 64-Core Processor
CPU family:                              25
Model:                                   48
Thread(s) per core:                      2
Core(s) per socket:                      64
Socket(s):                               1
Stepping:                                1
Frequency boost:                         enabled
CPU(s) scaling MHz:                      56%
CPU max MHz:                             3541.0149
CPU min MHz:                             1500.0000
BogoMIPS:                                3992.69
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm ibpb_exit_to_user
Virtualization:                          AMD-V
L1d cache:                               2 MiB (64 instances)
L1i cache:                               2 MiB (64 instances)
L2 cache:                                32 MiB (64 instances)
L3 cache:                                256 MiB (8 instances)
NUMA node(s):                            4
NUMA node0 CPU(s):                       0-15,64-79
NUMA node1 CPU(s):                       16-31,80-95
NUMA node2 CPU(s):                       32-47,96-111
NUMA node3 CPU(s):                       48-63,112-127
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; Safe RET
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable: No microcode
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] numpy==2.4.2
[pip3] pytorch-triton-rocm==3.5.1
[pip3] torch==2.9.1+rocm6.4
[pip3] torchaudio==2.9.1+rocm6.4
[pip3] torchvision==0.24.1+rocm6.4
[conda] Could not collect
```

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @jianyuh @mruberry @walterddr @xwang233 @Lezcano

### PR DESCRIPTION
**Pull request by naromero77amd**
Fixes #182952.

This PR changes the preferred backend for the fp16 backward GEMMs for ROCm only:
- Routes ROCm fp16 backward GEMMs from hipBLASLt to rocBLAS so backward subnormals are preserved.
- Keeps forward fp16 GEMMs on hipBLASLt while restoring `ROCmBackwardPassGuard` behavior for backward fp16 GEMMs.
- Adds matmul coverage for the ROCm fp16 subnormal gradient behavior.

Made with [Cursor](https://cursor.com)

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @pragupta @jerrymannil @xinyazhang

### CODE REVIEW

**Comment by naromero77amd** [comment]
CI artifact check for the new ROCm fp16 subnormal backward matmul test:

MI200 (`ciflow/rocm-mi200`, run `25944054660`, head `2b13217d14e4cf2d836bfaf05a950d0cb0c4fd2c`): the new test ran in shard `linux-jammy-rocm-py3.10-mi200 / test (default, 2, 6, linux.rocm.gpu.2)` via artifact `test-reports-runattempt1-test-default-2-6-linux.rocm.gpu.2_76271514823.zip`. All four parametrizations passed:

- `test_fp16_backward_preserves_subnormals_rocm_batched_False_backend_cublas_cuda`
- `test_fp16_backward_preserves_subnormals_rocm_batched_False_backend_cublaslt_cuda`
- `test_fp16_backward_preserves_subnormals_rocm_batched_True_backend_cublas_cuda`
- `test_fp16_backward_preserves_subnormals_rocm_batched_True_backend_cublaslt_cuda`

MI300 (`ciflow/rocm-mi300`, run `25944054269`, same head): verified from the test reports that the new test did not run on MI300. It was collected in the `linux-noble-rocm-py3.12-mi300 / test (default, 5, 6, linux.rocm.gpu.gfx942.1)` shard and all four parametrizations were skipped with `skipIfRocm: test only runs on ('gfx90a',)`.

**Review by jeffdaily — APPROVED** [review]

**Comment by jeffdaily** [comment]
@claude Analyze CI failures
